Added request body buffer regeneration

dylandreimerink opened the pull request:
I made a fix for issue #41 and #42.
Optional support for body buffer regeneration.

bungle commented on lib/resty/upload.lua:
```diff
@@ -93,21 +111,37 @@ function _M.set_timeout(self, timeout)
+end
+
+local function finish_body(self)
+    if self.resore_body then
```
typo here

bungle commented on lib/resty/upload.lua:
```diff
@@ -62,20 +65,35 @@ function _M.new(self, chunk_size, max_line_size)
         sock = sock,
         size = chunk_size or CHUNK_SIZE,
         line_size = max_line_size or MAX_LINE_SIZE,
+        resore_body = restore_body_buffer or false,
```
`resore` -> `restore`? Also `or false` not needed, though, doesn't hurt much either.

bungle commented on lib/resty/upload.lua:
```diff
@@ -145,6 +181,8 @@ local function read_body_part(self)
 
         local data = sock:receive(2)
         if data == "--" then
+            append_body(self, "\r\n--" .. self.boundary .. "--\r\n")
```
We always concatenate strings here. Even when `restore_body` is `false`. Same thing happens in other `append_body` places. Maybe just checking the boolean variable doesn't warrant its own function and that can be inlined. Then we also solve this issue.

bungle commented on lib/resty/upload.lua:
```diff
@@ -93,21 +111,37 @@ function _M.set_timeout(self, timeout)
 
     local line, err = read_line(self.line_size)
     if not line then
+        finish_body(self);
```
semicon `;` not needed.

dylandreimerink commented:
I fixed the typo and implemented you suggestion. I am not sure though if it is faster to append multiple times like I have done now or to create a new string and append once.

bungle commented:
Yes, previously you created strings even when not appending anything.

bungle commented on lib/resty/upload.lua:
```diff
@@ -93,21 +111,39 @@ function _M.set_timeout(self, timeout)
     return sock:settimeout(timeout)
 end
 
+local function append_body(self, ...)
```
I don't think you need to use varargs here, as you only ever have at most 3 args. Also you don't need to create a table with varargs, you can `select("#", ...)` and `select(1, ...)`.

bungle commented:
@dylandreimerink thanks for working on this. I think that one remaining thing is to add test(s).

dylandreimerink commented:
I took a look at the source code for [ngx.req.append_body](https://github.com/openresty/lua-nginx-module/blob/master/src/ngx_http_lua_req_body.c#L614) which is quite big and does a memcpy. So I think it is best to ditch the whole idea of a separate function ( like u suggested before ) and concatenate a single string in the main functions. Then calling ngx.req.append_body once right before ngx.req.finish_body to minimize overhead.

I will work on that as soon as I have time

dylandreimerink commented:
This should do it. I replaced the appending by a [string buffer](https://www.lua.org/pil/11.6.html) which should make it more performant. If this passes the only thing left to do is to make tests. I have to admit however that I do not yet understand how the this unit test framework works, so i may need a little bit of guidance.

dylandreimerink commented:
Working from the assumption that append_body is relatively speaking slow since it has to copy data from the lua vm to nginx the idea is to data in lua for as long as possible. Since lua strings are immutable they have to be recreated and copied every time you want to concatenate to an existing string. The solution would be to have a native implementation of a mutable string or string buffer. But since I have little hope for that in the near future I think the use of a table to store a list of strings and concatenating them once is the best option.

The reason I am going to these lengths is because I will be using this feature in proxy with a high throughput, so for me personally speed and optimization is quite important. So in light of that I will benchmark the different solutions so we can make an informed decision instead of having to work based on assumptions.

dylandreimerink commented:
I made the [benchmark](https://github.com/dylandreimerink/lua-resty-upload-benchmark). The [results](https://github.com/dylandreimerink/lua-resty-upload-benchmark/blob/master/results.txt) are to me a bit unexpected. I tested 4 variations and the original, the summarized results are as follows:

Name | total time | time per req (avg) | reqests per second(avg)
-- | -- | -- | --
Base test | 87.77 | 8.777 | 11393.38
complex string buffer test | 90.987 | 9.099 | 10990.54
simple string buffer test | 90.971 | 9.097 | 10992.56
string concat test | 90.843 | 9.084 | 11008.04
No buffer test | 90.233 | 9.023 | 11082.48

The variation without the string buffer is the fastest but the difference is almost negligible.
So I went with that solution.

dylandreimerink commented:
@bungle Have you had time to look at the last changes?

bungle commented on lib/resty/upload.lua:
```diff
@@ -62,20 +65,35 @@ function _M.new(self, chunk_size, max_line_size)
     if not read_line then
+
+        if restore_body_buffer then
+            ngx_finish_body()
```
The previous read, on line 72, was successful, but we didn't write the request body back, is this intended? Maybe there are many similar cases in code, but no easy way to solve? E.g. on errors we don't write the request body back.

dylandreimerink replied:
The sock:receiveuntil returns an iterator Lua function and doesn't actually read data from the buffer so there is noting to restore.

bungle commented on lib/resty/upload.lua:
```diff
@@ -93,21 +111,33 @@ function _M.set_timeout(self, timeout)
     end
 
+    if self.restore_body_buffer then
+        ngx_append_body(line .. "\r\n")
```
I think this is fine, but you could also have:
```lua
ngx_append_body(line)
ngx_append_body("\r\n")
```

dylandreimerink replied:
I don't see any point in not concatenating the string here, the same goes for the rest of the remarks regarding concatenation.

bungle commented on lib/resty/upload.lua:
```diff
@@ -145,6 +181,10 @@ local function read_body_part(self)
         local data = sock:receive(2)
         if data == "--" then
+            if self.restore_body_buffer then
+                ngx_append_body("\r\n--" .. self.boundary .. "--\r\n")
```
Same here:
```lua
ngx_append_body("\r\n--")
ngx_append_body(self.boundary)
ngx_append_body(ngx_append_body("--\r\n"))
```

Not a big deal.

bungle commented on lib/resty/upload.lua:
```diff
@@ -161,10 +201,18 @@ local function read_body_part(self)
         end
 
+        if self.restore_body_buffer then
+            ngx_append_body("\r\n--" .. self.boundary .. "\r\n")
```
This coud be cached before the while loop:

```lua
local boundary_delimiter = "\r\n--" .. self.boundary .. "\r\n"
```

And then here:
```lua
ngx_append_body(boundary_delimiter)
```

bungle commented on lib/resty/upload.lua:
```diff
@@ -232,7 +296,14 @@ local function read_preamble(self)
         local preamble = read2boundary(size)
         if not preamble then
+            if self.restore_body_buffer then
+                ngx_append_body("--" .. self.boundary)
```
Ditto here:
```lua
ngx_append_body("--")
ngx_append_body(self.boundary)
```

Yes, bikeshedding :-).

bungle commented on lib/resty/upload.lua:
```diff
@@ -203,7 +264,10 @@ local function read_header(self)
+local function eof(self)
+    if self.restore_body_buffer then
+        ngx_finish_body()
+    end
```
new line after this.

bungle commented on lib/resty/upload.lua:
```diff
@@ -232,7 +296,14 @@ local function read_preamble(self)
         if not preamble then
+            if self.restore_body_buffer then
+                ngx_append_body("--" .. self.boundary)
+            end
```
new line after this.

bungle commented on lib/resty/upload.lua:
```diff
@@ -232,7 +296,14 @@ local function read_preamble(self)
+            end
             break
+        else
+            if self.restore_body_buffer then
```
You can combine `else` and `if` here to `elseif`.

bungle commented on lib/resty/upload.lua:
```diff
@@ -232,7 +296,14 @@ local function read_preamble(self)
+                ngx_append_body("--" .. self.boundary)
+            end
             break
+        else
```
new line before this.

bungle commented on lib/resty/upload.lua:
```diff
@@ -174,15 +222,28 @@ local function read_header(self)
     if err then
+        if self.restore_body_buffer then
+            ngx_finish_body()
+        end
```
new line after this.

bungle commented on lib/resty/upload.lua:
```diff
@@ -174,15 +222,28 @@ local function read_header(self)
     if dummy then
+        if self.restore_body_buffer then
+            ngx_finish_body()
+        end
```
new line after this.

bungle commented on lib/resty/upload.lua:
```diff
@@ -174,15 +222,28 @@ local function read_header(self)
     local dummy, err = read_line(1)
     if dummy then
+        if self.restore_body_buffer then
+            ngx_finish_body()
```
do we also need to restore that `dummy`?

dylandreimerink replied:
The dummy is there to detect if the line we are trying to read exceeded the size of self.line_size. When this happens the request can't be processed further. This means that the remainder of the request will not be restored anyway. So there is no point is appending the one byte.

bungle commented on lib/resty/upload.lua:
```diff
@@ -93,21 +111,33 @@ function _M.set_timeout(self, timeout)
     if not line then
+        if self.restore_body_buffer then
+            ngx_finish_body()
+        end
```
new line after this.

bungle commented on lib/resty/upload.lua:
```diff
@@ -62,20 +65,35 @@ function _M.new(self, chunk_size, max_line_size)
+
     local read2boundary, err = sock:receiveuntil("--" .. boundary)
     if not read2boundary then
+
```
this new line is not needed.

bungle commented on lib/resty/upload.lua:
```diff
@@ -62,20 +65,35 @@ function _M.new(self, chunk_size, max_line_size)
         sock = sock,
         size = chunk_size or CHUNK_SIZE,
         line_size = max_line_size or MAX_LINE_SIZE,
+        restore_body_buffer = restore_body_buffer or false,
```
`or false` not strictly needed, as this is only for internal use.

bungle commented on lib/resty/upload.lua:
```diff
@@ -93,21 +111,33 @@ function _M.set_timeout(self, timeout)
     return sock:settimeout(timeout)
 end
 
-
```
restore this new line.

bungle commented on lib/resty/upload.lua:
```diff
@@ -93,21 +111,33 @@ function _M.set_timeout(self, timeout)
     if dummy then
+        if self.restore_body_buffer then
+            ngx_finish_body()
+        end
```
new line after this.

bungle commented on lib/resty/upload.lua:
```diff
@@ -93,21 +111,33 @@ function _M.set_timeout(self, timeout)
     if err then
+        if self.restore_body_buffer then
+            ngx_finish_body()
+        end
```
new line after this.

bungle commented on lib/resty/upload.lua:
```diff
@@ -122,6 +152,9 @@ local function discard_rest(self)
         if err and err ~= 'closed' then
+            if self.restore_body_buffer then
+                ngx_finish_body()
+            end
```
new line after this.

bungle commented on lib/resty/upload.lua:
```diff
@@ -137,6 +170,9 @@ local function read_body_part(self)
     if err then
+        if self.restore_body_buffer then
+            ngx_finish_body()
+        end
```
new line after this.

bungle commented on lib/resty/upload.lua:
```diff
@@ -174,15 +222,28 @@ local function read_header(self)
     if err then
+        if self.restore_body_buffer then
+            ngx_finish_body()
+        end
```
new line after this.
